src/libfetchers/union-input-accessor.cc

```diff
@@ -0,0 +1,80 @@
+#include "union-input-accessor.hh"
+
+namespace nix {
+
+struct UnionInputAccessor : InputAccessor
+{
+    std::map<CanonPath, ref<InputAccessor>> mounts;
+
+    UnionInputAccessor(std::map<CanonPath, ref<InputAccessor>> _mounts)
+        : mounts(std::move(_mounts))
+    {
+        // Currently we require a root filesystem. This could be relaxed.
+        assert(mounts.contains(CanonPath::root));
+
+        // FIXME: should check that every mount point exists. Or we
+        // could return dummy parent directories automatically.
```

Comment:
```suggestion
        // FIXME: return dummy parent directories automatically?
```

I think it's better to do it lazily (regardless of the other decision). They're lazy trees after all.

I'd be happy to unconditionally and lazily generate those directories, even if they shadow pre-existing files.
If this were a primop, a user library could be written to add checking for such "conflicts", but avoiding a built-in check is more difficult and less lazy. But as mentioned, I don't think it's even worth checking, so such a user library is just a last resort.